src/expected_loglik.jl

```diff
@@ -0,0 +1,151 @@
+abstract type QuadratureMethod end
+struct DefaultQuadrature <: QuadratureMethod end
+struct Analytic <: QuadratureMethod end
+
+struct GaussHermite <: QuadratureMethod
+    n_points::Int
+end
+GaussHermite() = GaussHermite(20)
+
+struct MonteCarlo <: QuadratureMethod
+    n_samples::Int
+end
+MonteCarlo() = MonteCarlo(20)
+
+_default_quadrature(_) = GaussHermite()
+
+"""
+    expected_loglik(quadrature::QuadratureMethod, y::AbstractVector, q_f::AbstractVector{<:Normal}, lik)
+
+This function computes the expected log likelihood:
+
+```math
+    ∫ q(f) log p(y | f) df
+```
+where `p(y | f)` is the process likelihood.
+
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)). The marginal distributions of `q(f)` are given by `q_f`.
+
+`quadrature` determines which method is used to calculate the expected log
+likelihood - see [`elbo`](@ref) for more details.
+
+# Extended help
+
+`q(f)` is assumed to be an `MvNormal` distribution and `p(y | f)` is assumed to
+have independent marginals such that only the marginals of `q(f)` are required.
+"""
+expected_loglik(quadrature, y, q_f, lik)
+
+"""
+    expected_loglik(::DefaultQuadrature, y::AbstractVector, q_f::AbstractVector{<:Normal}, lik)
+
+The expected log likelihood.
+Defaults to a closed form solution if it exists, otherwise defaults to
+Gauss-Hermite quadrature.
+"""
+function expected_loglik(
+    ::DefaultQuadrature,
+    y::AbstractVector,
+    q_f::AbstractVector{<:Normal},
+    lik,
+)
+    quadrature = _default_quadrature(lik)
+    return expected_loglik(quadrature, y, q_f, lik)
+end
+
+function expected_loglik(
+    mc::MonteCarlo, y::AbstractVector, q_f::AbstractVector{<:Normal}, lik
+)
+    # take `n_samples` reparameterised samples
+    f_μ = mean.(q_f)
+    fs = f_μ .+ std.(q_f) .* randn(eltype(f_μ), length(q_f), mc.n_samples)
+    lls = loglikelihood.(lik.(fs), y)
+    return sum(lls) / mc.n_samples
+end
+
+function expected_loglik(
+    gh::GaussHermite,
+    y::AbstractVector,
+    q_f::AbstractVector{<:Normal},
+    lik,
+)
+    # Compute the expectation via Gauss-Hermite quadrature
+    # using a reparameterisation by change of variable
+    # (see e.g. en.wikipedia.org/wiki/Gauss%E2%80%93Hermite_quadrature)
+    xs, ws = gausshermite(gh.n_points)
+    # size(fs): (length(y), n_points)
+    fs = √2 * std.(q_f) .* transpose(xs) .+ mean.(q_f)
```

Comment:
Would
```suggestion
    fs = √2 * std.(q_f) .* xs' .+ mean.(q_f)
```
work as well?